### ISSUE TITLE
Very slow in kubernetes environment

### ISSUE THREAD

**Issue by KyleGobel** [aspnet/KestrelHttpServer#2219]
I have a problem I've been struggling with for a couple weeks now.

I run a couple pretty thin API dotnet core projects in kubernetes.  Some of these (it almost seems hit or miss), are unbelievably slow to start up.  It takes normally more than 5 minutes before my first health check passes.

After that most requests will timeout for another couple minutes and finally maybe after 10 - 15 min it will finally start to run fine. 

This is most noticeable with database connections.  All database connection open calls will timeout for the first 10 minutes or so, after they've been warmed up or called a few times they will finally start to work.

At startup I will frequently get a bunch of these these messages:
```
[15:00:18 WRN] Heartbeat took longer than "00:00:01" at "12/13/2017 15:00:15 +00:00".
```

I could use some debugging help, are there any hunches of what the problem **could** be?  Is there anything I can and should be checking?  Any logs or metrics I can provide that might narrow down my problem?

**Environment**
---------
 - Kubernetes cluster running on AWS provisioned with kops.
- 1x c4.large master 
- 3x c4.xlarge instance sizes 
- base docker image of aspnetcore:2.0.3.  Problem exists in both jessie and stretch variants.  Also tried custom ubuntu image (though basically copied from the dotnet-runtime docker images)

Not memory/cpu/network constrained as far as I can tell.  I've tried on new nodes and also new clusters

**Comment by muratg** [comment]
@KyleGobel Does your app start much faster in your dev environment? (I assume so...)   Do you do any specific expensive operations in your startup?  Did you try getting logs to see what may be taking time?  

cc @natemcmaster

**Comment by KyleGobel** [comment]
@muratg Yes everything is incredibly fast in the develop environment.  Even just taking my docker image and running it locally has no problems what so ever.  I can only get it to misbehave when it's put into kubernetes.

I don't do any expensive operations that I know of in my startup.  I am attempting to clean up a minimal example again (I've done this a couple times) that I will share.

I will post more verbose logs shortly, but here was the last run with all/any custom middleware turned off.

```
[01:22:26 INF] App starting in Staging env
The path is: /app/
[01:25:30 DBG] Hosting starting
[01:25:39 WRN] Heartbeat took longer than "00:00:01" at "12/14/2017 01:25:37 +00:00".
[01:25:44 DBG] Hosting started
[01:25:44 DBG] Loaded hosting startup assembly TaskService.Api
Hosting environment: Staging
Content root path: /app
Now listening on: http://[::]:80
Application started. Press Ctrl+C to shut down.
[01:26:26 DBG] Connection id "0HLA2I6OJB703" started.
[01:26:36 DBG] Connection id "0HLA2I6OJB703" received FIN.
[01:26:36 DBG] Connection id "0HLA2I6OJB703" disconnecting.
[01:26:38 DBG] Connection id "0HLA2I6OJB703" sending FIN.
[01:26:40 DBG] Connection id "0HLA2I6OJB704" started.
[01:26:40 DBG] Connection id "0HLA2I6OJB703" stopped.
[01:26:45 INF] Request starting HTTP/1.1 GET http://100.96.2.169:80/metrics  
[01:26:50 DBG] Connection id "0HLA2I6OJB704" received FIN.
[01:26:51 DBG] Connection id "0HLA2I6OJB704" disconnecting.
[01:26:51 DBG] Connection id "0HLA2I6OJB704" sending FIN.
[01:27:12 DBG] Request did not match any routes.
[01:27:13 INF] Request finished in 29577.0212ms 0 
[01:27:14 DBG] Connection id "0HLA2I6OJB704" stopped. 
```

I normally have some custom prometheus middleware running that provides the /metrics endpoint that gets pinged every 5 sec, but in this case that would have been a 404 that took about 30 sec.

There is a pretty long gap between my first log message immediately after configure logging and Hosting Starting.  I'm not sure what is going on during that time.

**Comment by muratg** [comment]
That's indeed a long time.  Are you able to try it on another Kubernetes setting at all?  I don't think Kubernetes would be the reason but just in case if it's reasonably easy to do. 

cc @Tratcher in case he has any thoughts.

**Comment by muratg** [comment]
@KyleGobel The following line is interesting:
`[01:25:44 DBG] Loaded hosting startup assembly TaskService.Api`

Is TaskService.Api.dll an assembly your app uses?  Looks like you're using hosting lightup...  Is this by design, or accidental, I wonder.

**Comment by KyleGobel** [comment]
``TaskService.Api`` is the name of the project/app.  If that is not normal it was definitely accidental.  I'm not trying to do anything tricky or clever here.

**Comment by KyleGobel** [comment]
Ok, well in trying to set up an example I think I found the reasoning.

On certain containers I had monitored their CPU and setup Kubernetes CPU requests and limits for what these containers run at (I had set them pretty low because some of these get very little traffic and the average cpu usage was nearly nothing).  It turns out that if you limit their CPU too much they take longer to start and get going (duh)🤦‍♂️.  I didn't realize (or think) about how much horsepower they use at startup.

I also didn't realize that unlike memory, kubernetes doesn't simply turn off/restart the container if it goes over it's CPU limit, it can instead just throttles your container limited cycles.

My bad! @muratg thanks for trying to help me out!  Really appreciate it.
